File: scripts_fetch_idea_details.py
import re,json,requests
from pathlib import Path

text=Path(r'd:\work\gray\src\data\siteContent.js').read_text(encoding='utf-8')
# isolate ideasDirectory literal
m=re.search(r'export const ideasDirectory = \[(.*?)\]\.map\(', text, re.S)
chunk=m.group(1) if m else ''
obj_matches=re.findall(r'\{(.*?)\},', chunk, re.S)
items=[]
for body in obj_matches:
    def get_field(name):
        m=re.search(rf"{name}:\s*'([^']*)'", body)
        if m: return m.group(1)
        m=re.search(rf'{name}:\s*"([^"]*)"', body)
        if m: return m.group(1)
        return None
    title=get_field('title'); client=get_field('client'); desc=get_field('description'); url=get_field('url')
    if title and url:
        slug=re.sub(r'[^a-z0-9]+','-',title.lower()).strip('-')
        items.append({'slug':slug,'title':title,'client':client,'description':desc,'url':url})

session=requests.Session()

def flatten_text(node,out):
    if isinstance(node,dict):
        if node.get('nodeType')=='text' and isinstance(node.get('value'),str):
            v=node['value'].strip()
            if v: out.append(v)
        for v in node.values(): flatten_text(v,out)
    elif isinstance(node,list):
        for i in node: flatten_text(i,out)

res={}
for item in items:
    rec={**item,'originTitle':None,'originHeadline':None,'originParagraphs':[],'originMedia':[]}
    try:
        html=session.get(item['url'],timeout=20).text
        m=re.search(r'<script id="__NEXT_DATA__" type="application/json">(.*?)</script>',html,re.S)
        if m:
            data=json.loads(m.group(1))
            page=((data.get('props') or {}).get('pageProps') or {}).get('data',{}).get('page',{})
            rec['originTitle']=page.get('title')
            rec['originHeadline']=page.get('headline')
            consys=(page.get('consysCollection') or {}).get('items') or []
            paras=[]
            medias=[]
            def walk(o):
                if isinstance(o,dict):
                    if o.get('contentType') and isinstance(o.get('url'),str) and 'ctfassets.net' in o['url']:
                        medias.append({'url':o['url'],'contentType':o.get('contentType'),'title':o.get('title')})
                    if 'json' in o and isinstance(o['json'],dict):
                        bag=[]; flatten_text(o['json'],bag)
                        t=' '.join(bag).strip()
                        if t: paras.append(t)
                    for v in o.values(): walk(v)
                elif isinstance(o,list):
                    for i in o: walk(i)
            walk(consys)
            # unique keep order
            p2=[]; seen=set()
            for p in paras:
                if p not in seen: seen.add(p); p2.append(p)
            m2=[]; seen=set()
            for med in medias:
                if med['url'] not in seen: seen.add(med['url']); m2.append(med)
            rec['originParagraphs']=p2[:24]
            rec['originMedia']=m2[:16]
    except Exception:
        pass
    res[item['slug']]=rec

Path(r'd:\work\gray\src\data\ideaDetails.js').write_text('export const ideaDetailsBySlug = '+json.dumps(res,ensure_ascii=False,indent=2)+'\n',encoding='utf-8')
print('items',len(items),'details',len(res))

File: scripts_fetch_idea_details.mjs
import { writeFileSync } from 'node:fs'
import { ideasDirectory } from './src/data/siteContent.js'

function extractText(node, bag) {
  if (Array.isArray(node)) {
    for (const n of node) extractText(n, bag)
    return
  }
  if (!node || typeof node !== 'object') return
  if (node.nodeType === 'text' && typeof node.value === 'string') {
    const t = node.value.trim()
    if (t) bag.push(t)
  }
  for (const v of Object.values(node)) extractText(v, bag)
}

function walk(obj, paragraphs, media) {
  if (Array.isArray(obj)) {
    for (const x of obj) walk(x, paragraphs, media)
    return
  }
  if (!obj || typeof obj !== 'object') return

  if (obj.contentType && obj.url && typeof obj.url === 'string' && obj.url.includes('ctfassets.net')) {
    media.push({ url: obj.url, contentType: obj.contentType, title: obj.title ?? null })
  }

  if (obj.json && typeof obj.json === 'object') {
    const bag = []
    extractText(obj.json, bag)
    const t = bag.join(' ').trim()
    if (t) paragraphs.push(t)
  }

  for (const v of Object.values(obj)) walk(v, paragraphs, media)
}

const details = {}
for (const item of ideasDirectory) {
  const rec = {
    ...item,
    originTitle: null,
    originHeadline: null,
    originParagraphs: [],
    originMedia: [],
  }
  try {
    const res = await fetch(item.url, { redirect: 'follow' })
    const html = await res.text()
    const m = html.match(/<script id="__NEXT_DATA__" type="application\/json">([\s\S]*?)<\/script>/)
    if (m) {
      const data = JSON.parse(m[1])
      const page = data?.props?.pageProps?.data?.page ?? {}
      rec.originTitle = page?.title ?? null
      rec.originHeadline = page?.headline ?? null
      const paragraphs = []
      const media = []
      walk(page?.consysCollection?.items ?? [], paragraphs, media)
      rec.originParagraphs = [...new Set(paragraphs)].slice(0, 24)
      const seen = new Set()
      rec.originMedia = media.filter((x) => (seen.has(x.url) ? false : (seen.add(x.url), true))).slice(0, 16)
    }
  } catch {}
  details[item.slug] = rec
}

writeFileSync('./src/data/ideaDetails.js', `export const ideaDetailsBySlug = ${JSON.stringify(details, null, 2)}\n`, 'utf8')
console.log('generated', Object.keys(details).length)
